File: src/puppypi_adapter/puppypi_adapter/mode_adapter_node.py
"""Mode Adapter Node: manages PuppyPi posture mode transitions.

Handles posture commands: STAND, SIT, LIE_DOWN, RECOVER, FREEZE
(gaijin2.md section 7.2)

P0-3: use_sim=True（默认）时只记录姿态命令不执行；
      use_sim=False 时尝试导入 PuppyPi SDK，失败则报错退出。
"""
import rclpy
from rclpy.node import Node
from std_msgs.msg import String
from std_msgs.msg import Bool


class ModeAdapterNode(Node):
    """Manages PuppyPi posture and mode transitions."""

    POSTURES = ['STAND', 'SIT', 'LIE_DOWN', 'RECOVER', 'FREEZE']

    def __init__(self):
        super().__init__('mode_adapter')

        # P0-3: 仿真/真机模式开关
        self.declare_parameter('use_sim', True)
        self.declare_parameter('backend', '')
        self.use_sim = self.get_parameter('use_sim').value
        configured_backend = str(self.get_parameter('backend').value or '').strip().lower()

        self.current_posture = 'UNKNOWN'
        self.motion_enabled = False

        # Publishers / Subscribers
        self.create_subscription(String, '/robot/posture_cmd', self._on_posture_cmd, 10)
        self.create_subscription(Bool, '/robot/motion_enable', self._on_enable, 10)

        # HardwareInterface handles both sim and real hardware abstraction
        self.hw = None
        self.sdk_connected = False
        # P0-4: backend_actual 记录"实际生效"的后端, 绝不因回退而谎报 configured 后端
        # use_sim=True -> backend 'sim' (外部仿真器拥有动力学, 本节点只发命令与
        # 使能/急停; 传感器数据由仿真器经 ROS topic 注入, 见 sim_interface.py)。
        # 这与 'mock'(进程内自成体系的假动力学) 语义不同, 不得混用。
        self.backend_requested = configured_backend or ('sim' if self.use_sim else 'real')
        self.backend_actual = None

        try:
            from .hardware_interface import HardwareInterface
            self.hw = HardwareInterface.create({'backend': self.backend_requested})
            self.sdk_connected = self.hw.initialize()
            self.backend_actual = self.backend_requested
            if not self.sdk_connected:
                # P0-4: 初始化失败不再静默回退 mock —— 显式报错并退出, 避免
                # "配置 real 却跑在 mock 上" 的静默降级被误当作真实运行.
                raise RuntimeError(
                    f'HardwareInterface ({self.backend_requested}) initialize() returned False')
        except Exception as exc:
            self.get_logger().fatal(
                f'HardwareInterface ({self.backend_requested}) unavailable: {exc}. '
                f'P0-4 禁止静默回退 mock: 请修正 backend 配置或安装对应后端 '
                f'(sim 需 sim_interface, real 需 PuppyPi SDK 且预检通过).'
            )
            raise

        self.backend_actual = self.backend_requested
        mode_tag = ('[SIM]' if self.use_sim
                    else ('[HARDWARE]' if self.sdk_connected else '[UNKNOWN]'))
        self.get_logger().info(
            f'Mode adapter started {mode_tag} '
            f'(backend={self.backend_actual}, sdk_connected={self.sdk_connected})')

    def _on_posture_cmd(self, msg: String):
        """Handle posture command."""
        posture = msg.data.upper()
        if posture not in self.POSTURES:
            self.get_logger().warn(f'Unknown posture: {posture}')
            return
        self.get_logger().info(f'Posture command: {posture}')
        self.current_posture = posture

        if self.hw:
            if posture == 'FREEZE':
                self.hw.emergency_stop('posture_cmd_freeze')
            elif posture == 'RECOVER':
                self.hw.release_emergency_stop()
                self.hw.enable_motors(True)
            elif posture == 'STAND':
                self.hw.enable_motors(True)
            elif posture in ('SIT', 'LIE_DOWN'):
                self.hw.send_velocity(0.0, 0.0, 0.0)

    def _on_enable(self, msg: Bool):
        """Handle motor enable/disable."""
        self.motion_enabled = msg.data
        self.get_logger().info(f'Motors {"enabled" if msg.data else "disabled"}')
        if self.hw:
            self.hw.enable_motors(msg.data)


def main(args=None):
    rclpy.init(args=args)
    node = ModeAdapterNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    node.destroy_node()
    rclpy.shutdown()

File: cpp_src/puppypi_adapter/mode_adapter_node.cpp
// Mode Adapter Node 实现
//
// 对应 Python: puppypi_adapter/mode_adapter_node.py
// ROS2 节点，包含 main()
#include "puppypi_adapter/mode_adapter_node.h"

#include <algorithm>
#include <cctype>
#include <stdexcept>

namespace puppypi_adapter {

// 支持的姿态列表
const std::vector<std::string> ModeAdapterNode::POSTURES = {
    "STAND", "SIT", "LIE_DOWN", "RECOVER", "FREEZE",
};

ModeAdapterNode::ModeAdapterNode()
    : rclcpp::Node("mode_adapter") {
    // 参数声明
    this->declare_parameter("use_sim", true);
    use_sim_ = this->get_parameter("use_sim").as_bool();

    // 订阅器
    posture_sub_ = this->create_subscription<std_msgs::msg::String>(
        "/robot/posture_cmd", 10,
        std::bind(&ModeAdapterNode::onPostureCmd, this,
                  std::placeholders::_1));
    enable_sub_ = this->create_subscription<std_msgs::msg::Bool>(
        "/robot/motion_enable", 10,
        std::bind(&ModeAdapterNode::onEnable, this,
                  std::placeholders::_1));

    // P0-3: SDK 初始化
    // use_sim=False 时尝试导入 PuppyPi SDK，当前未集成则报错退出
    if (!use_sim_) {
        RCLCPP_FATAL(this->get_logger(),
                     "use_sim=False but PuppyPi SDK unavailable. "
                     "Set use_sim:=true for simulation.");
        throw std::runtime_error("PuppyPi SDK not yet integrated");
    }

    const char* mode_tag = use_sim_ ? "[SIM]" : "[HARDWARE]";
    RCLCPP_INFO(this->get_logger(), "Mode adapter started %s", mode_tag);
}

void ModeAdapterNode::onPostureCmd(
    const std_msgs::msg::String::SharedPtr msg) {
    // 转大写
    std::string posture = msg->data;
    std::transform(posture.begin(), posture.end(), posture.begin(),
                   [](unsigned char c) { return std::toupper(c); });

    // 校验姿态
    bool valid = false;
    for (const auto& p : POSTURES) {
        if (p == posture) {
            valid = true;
            break;
        }
    }
    if (!valid) {
        RCLCPP_WARN(this->get_logger(), "Unknown posture: %s",
                    posture.c_str());
        return;
    }
    RCLCPP_INFO(this->get_logger(), "Posture command: %s",
                posture.c_str());
    // P0-3: 仿真模式只记录，真机模式调用 SDK
    current_posture_ = posture;
    // TODO: 真机模式调用 self.puppypi.set_posture(posture)
}

void ModeAdapterNode::onEnable(
    const std_msgs::msg::Bool::SharedPtr msg) {
    motion_enabled_ = msg->data;
    RCLCPP_INFO(this->get_logger(), "Motors %s",
                msg->data ? "enabled" : "disabled");
    // P0-3: 仿真模式只记录，真机模式调用 SDK
    // TODO: 真机模式调用 self.puppypi.enable_motors(msg->data)
}

}  // namespace puppypi_adapter

// ===== main =====
int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    try {
        rclcpp::spin(std::make_shared<puppypi_adapter::ModeAdapterNode>());
    } catch (const std::exception& e) {
        // 构造失败时 rclcpp::init 已执行，需要 shutdown
        rclcpp::shutdown();
        return 1;
    } catch (...) {
        rclcpp::shutdown();
        return 1;
    }
    rclcpp::shutdown();
    return 0;
}
